tower/src/util/map_future.rs

```diff
@@ -0,0 +1,105 @@
+use std::{
+    fmt,
+    future::Future,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// [`Service`] returned by the [`map_future`] combinator.
+///
+/// [`map_future`]: crate::util::ServiceExt::map_future
+#[derive(Clone)]
+pub struct MapFuture<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> MapFuture<S, F> {
+    /// Creates a new [`MapFuture`] service.
+    pub fn new(inner: S, f: F) -> Self {
+        Self { inner, f }
+    }
+
+    /// Returns a new [`Layer`] that produces [`MapFuture`] services.
+    ///
+    /// This is a convenience function that simply calls [`MapFutureLayer::new`].
+    ///
+    /// [`Layer`]: tower_layer::Layer
+    pub fn layer(f: F) -> MapFutureLayer<F> {
+        MapFutureLayer::new(f)
+    }
+
+    /// Get a reference to the inner service
+    pub fn get_ref(&self) -> &S {
+        &self.inner
+    }
+
+    /// Get a mutable reference to the inner service
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.inner
+    }
+
+    /// Consume `self`, returning the inner service
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
+}
+
+impl<R, S, F, T, E, Fut> Service<R> for MapFuture<S, F>
+where
+    S: Service<R>,
+    F: FnMut(S::Future) -> Fut,
+    E: From<S::Error>,
+    Fut: Future<Output = Result<T, E>>,
+{
+    type Response = T;
+    type Error = E;
```

Comment:
TIOLI: we could reduce the number of type parameters here using `TryFuture`:
```suggestion
impl<R, S, F, Fut> Service<R> for MapFuture<S, F>
where
    S: Service<R>,
    F: FnMut(S::Future) -> Fut,
    Fut: TryFuture,
    Fut::Error: From<S::Error>,
{
    type Response = Fut::Ok;
    type Error = Fut::Error;
```
not a big deal though...

Reply:
Interesting! How does this work in terms of not exposing things from the `futures` crate?